featurebyte/models/task.py

```diff
@@ -2,16 +2,90 @@
 Task document model
 """
 
+from collections import defaultdict
 from datetime import datetime
 from typing import Any, Dict, List, Optional
 from uuid import UUID, uuid4
 
 import pymongo
-from pydantic import Field
+from pydantic import BaseModel, Field
 
 from featurebyte.models.base import FeatureByteBaseDocumentModel
 
 
+class LogMessage(BaseModel):
+    """Log message schema"""
+
+    percent: int
+    message: Optional[str]
+
+
+class ProgressHistory(BaseModel):
+    """
+    Progress history schema
+    """
+
+    data: List[LogMessage] = Field(default=[])
+    # compress_at is the last percent at which compression was done
+    # if compress_at is 0, it means no compression has been done
+    # if compress_at is 100, it means all messages have been compressed
+    compress_at: int = Field(default=0)
+
+    def compress(self, max_messages: int) -> "ProgressHistory":
+        """
+        Compress progress history
+
+        Parameters
+        ----------
+        max_messages : int
+            Maximum number of messages to keep
+
+        Returns
+        -------
+        ProgressHistory
+            Compressed progress history
+        """
+        if len(self.data) <= max_messages:
+            return self
+
+        # Step 1: Map percents to message indices
+        percent_to_indices = defaultdict(list)
+        for idx, msg in enumerate(self.data):
+            percent_to_indices[msg.percent].append(idx)
+
+        # Step 2: Initialize variables
+        compressed_indices = set()
+        total_messages = len(self.data)
+        current_percent = self.compress_at
+
+        # Step 3: Iterate over percents from compress_at upwards
+        for percent in range(self.compress_at, 101):
+            indices = percent_to_indices.get(percent, [])
+            if len(indices) > 1:
+                # Keep only the last message, remove the rest
+                indices_to_remove = indices[:-1]
+                compressed_indices.update(indices_to_remove)
+                total_messages -= len(indices_to_remove)
+
+            # Update compress_at to the current percent
+            current_percent = percent
+
+            # Stop if total messages are within the limit
+            if total_messages <= max_messages:
```

Comment:
Is the intention here to only compress earlier messages when necessary?

Reply:
Yes, only compress it further if the limit is exceeded again.